File: frontend/android/app/src/main/java/com/fridgeboard/app/DeepLinkPlugin.java
package com.fridgeboard.app;

import android.content.Intent;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "DeepLink")
public class DeepLinkPlugin extends Plugin {
    private static DeepLinkPlugin instance;
    private static String pendingUrl;
    private String initialUrl;

    @Override
    public void load() {
        instance = this;
        Intent intent = getActivity().getIntent();
        initialUrl = intent == null || intent.getData() == null ? null : intent.getDataString();
    }

    @PluginMethod
    public void getInitialUrl(PluginCall call) {
        String url = initialUrl != null ? initialUrl : pendingUrl;
        initialUrl = null;
        pendingUrl = null;
        JSObject result = new JSObject();
        result.put("url", url);
        call.resolve(result);
    }

    public static void receiveIntent(Intent intent) {
        if (intent == null || intent.getData() == null) return;
        String url = intent.getDataString();
        DeepLinkPlugin plugin = instance;
        if (plugin == null) {
            pendingUrl = url;
            return;
        }
        plugin.notifyUrl(url);
    }

    private void notifyUrl(String url) {
        JSObject event = new JSObject();
        event.put("url", url);
        notifyListeners("urlOpen", event);
    }
}

File: frontend/ios/App/App/DeepLinkPlugin.swift
import Capacitor
import Foundation

@objc(DeepLinkPlugin)
public class DeepLinkPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DeepLinkPlugin"
    public let jsName = "DeepLink"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getInitialUrl", returnType: CAPPluginReturnPromise)
    ]

    private static var pendingUrl: String?
    private static weak var instance: DeepLinkPlugin?

    override public func load() {
        Self.instance = self
    }

    @objc func getInitialUrl(_ call: CAPPluginCall) {
        let url = Self.pendingUrl
        Self.pendingUrl = nil
        call.resolve(["url": url ?? NSNull()])
    }

    public static func receive(url: URL) {
        let value = url.absoluteString
        pendingUrl = value
        guard let instance else { return }
        instance.notifyListeners("urlOpen", data: ["url": value])
    }
}
